subprojects/packagefiles/quickjs-ng/meson.build

```diff
@@ -0,0 +1,106 @@
+# DO not forget to update QJS_VERSION_MAJOR/QJS_VERSION_MINOR/QJS_VERSION_PATCH when upgrading version
+project('quickjs-ng', 'c', version: '0.8.0', default_options: ['c_std=c11'])
+
+
+cc = meson.get_compiler('c')
+dep_m = cc.find_library('m', required: false)
+dep_threads = dependency('threads')
+deps = [dep_m, dep_threads]
+
+msvc_compat = cc.get_id() == 'msvc' or cc.get_id() == 'clang-cl'
+
+add_project_arguments(
+  '-DQJS_VERSION_MAJOR=0',
+  '-DQJS_VERSION_MINOR=8',
+  '-DQJS_VERSION_PATCH=0',
+  language: 'c',
+)
+
+add_project_arguments(
+  cc.get_supported_arguments(
+    '-Wno-implicit-fallthrough',
+    '-Wno-sign-compare',
+    '-Wno-missing-field-initializers',
+    '-Wno-unused-parameter',
+    '-Wno-unused-but-set-variable',
+    '-Wno-array-bounds',
+    '-Wno-format-truncation',
+    '-funsigned-char',
+  ),
+  language: 'c',
+)
+
+if msvc_compat
+  add_project_arguments(
+    cc.get_supported_arguments(
+      '-Wno-unsafe-buffer-usage',
+      '-Wno-sign-conversion',
+      '-Wno-nonportable-system-include-path',
+      '-Wno-implicit-int-conversion',
+      '-Wno-shorten-64-to-32',
+      '-Wno-reserved-macro-identifier',
+      '-Wno-reserved-identifier',
+      '-Wdeprecated-declarations',
+      '/experimental:c11atomics',
+    ),
+    language: 'c',
+  )
+
+  add_project_arguments(
+    cc.get_supported_arguments(
+      '/wd4018',  # -Wno-sign-conversion
+      '/wd4061',  # -Wno-implicit-fallthrough
+      '/wd4100',  # -Wno-unused-parameter
+      '/wd4200',  # -Wno-zero-length-array
+      '/wd4242',  # -Wno-shorten-64-to-32
+      '/wd4244',  # -Wno-shorten-64-to-32
+      '/wd4245',  # -Wno-sign-compare
+      '/wd4267',  # -Wno-shorten-64-to-32
+      '/wd4388',  # -Wno-sign-compare
+      '/wd4389',  # -Wno-sign-compare
+      '/wd4710',  # Function not inlined
+      '/wd4711',  # Function was inlined
+      '/wd4820',  # Padding added after construct
+      '/wd4996',  # -Wdeprecated-declarations
+      '/wd5045',  # Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified
+    ),
+    language: 'c',
+  )
+
+endif
+
+if get_option('buildtype') == 'debug'
+  add_project_arguments(
+    cc.get_supported_arguments('-ggdb', '-fno-omit-frame-pointer'),
+    language: 'c',
+  )
+endif
+
+lib_src = files(
+  'cutils.c',
+  'libbf.c',
+  'libregexp.c',
+  'libunicode.c',
+  'quickjs-libc.c',
+  'quickjs.c',
+)
+
+c_defines = ['-D_GNU_SOURCE', '-DCONFIG_BIGNUM']
+
+if host_machine.system() == 'windows'
+  c_defines += ['-DWIN32_LEAN_AND_MEAN', '-D_WIN32_WINNT=0x0602']
+endif
+
+add_project_arguments(c_defines, language: 'c')
+
+# TODO: support shared libs
+quickjs_lib = static_library('quickjs', lib_src, dependencies: deps)
+
+quickjs_ng_dep = declare_dependency(
+  include_directories: include_directories('.'),
```

Comment:
I don't think this is a good idea. 
It exposes all headers, creating name collisions downstream.  
I encountered the same problem when adapting other projects like `tcc` or `mongoose`.  
The trick I used to avoid this, is like what I have done in here: https://github.com/lazy-eggplant/vs.http/tree/master/subprojects/packagefiles/mongoose
Basically I create a `meson-include` folder, and add symbolic links to the required headers in there.
However, this had a bug up until recently https://github.com/mesonbuild/meson/pull/14073 so it would not work with older versions :(.

Reply:
The only other alternatives I can see are
- to fork the project & better organize files so that we can have sane search paths
- keep the `.` and accepting things will break downstream when building, in scenarios which are quite obscure and hard to predict.
For context, a file `config.h` in tinycc was creating problems in one of my projects, but only on macos-14 with clang-19. I wasted way to many hours to figure that out.

Reply:
Yes, it is possible I guess. It makes the process of maintaining the patch more annoying than what it should be I guess. But it would work.